polkadot/node/collation-generation/src/lib.rs

```diff
@@ -598,3 +617,39 @@ fn erasure_root(
 	let chunks = polkadot_erasure_coding::obtain_chunks_v1(n_validators, &available_data)?;
 	Ok(polkadot_erasure_coding::branches(&chunks).root())
 }
+
+// Checks if the runtime supports `request_claim_queue` and executes it. Returns `Ok(None)`
+// otherwise. Any [`RuntimeApiError`]s are bubbled up to the caller.
+async fn fetch_claim_queue(
+	sender: &mut impl overseer::CollationGenerationSenderTrait,
+	relay_parent: Hash,
+) -> crate::error::Result<Option<BTreeMap<CoreIndex, VecDeque<ParasEntry<BlockNumber>>>>> {
+	if has_required_runtime(
+		sender,
+		relay_parent,
+		RuntimeApiRequest::CLAIM_QUEUE_RUNTIME_REQUIREMENT,
+	)
+	.await
+	{
+		let res = request_claim_queue(relay_parent, sender).await.await??;
+		Ok(Some(res))
+	} else {
+		gum::trace!(target: LOG_TARGET, "Runtime doesn't support `request_claim_queue`");
+		Ok(None)
+	}
+}
+
+fn fetch_scheduled_para(
+	claim_queue: &BTreeMap<CoreIndex, VecDeque<ParasEntry<BlockNumber>>>,
+	core_idx: CoreIndex,
+) -> Option<ScheduledCore> {
+	// TODO: In the future we will return more than one scheduled para_id hence the weird processing
+	// below
+	claim_queue
+		.get(&core_idx)?
+		.into_iter()
+		.map(|paras_entry| paras_entry.para_id())
+		.take(1)
+		.next()
+		.map(|para_id| ScheduledCore { para_id, collator: None })
+}
```

Comment:
This is an overkill if we just need one `ScheduledCore`. I'll fix it if we decide not to fetch more than one entry.

Reply:
Not relevant anymore